### PR TITLE
2238/add file download to research

### PR DESCRIPTION
**Pull request by AlfonsoGhislieri**
PR Checklist

- [x] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

This introduced the ability to upload files when creating and editing research articles as well as having the functionality to track the total number of downloads. 

This is PART 1, a lot of this functionality repeats with what already exists for the howtos. In order to avoid bloating this PR too much PART 2 will address refactoring, finally PART 3 will add icrementing the downloads via a firebase function to avoid unnecessary database calls.

## Git Issues

#2238

## Screenshots/Videos

image
---

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of a monthly dev call (first Monday of the month, open to all!).

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

### CODE REVIEW

File: src/setupTests.ts [hunk 1]
```diff
@@ -3,3 +3,5 @@
 // expect(element).toHaveTextContent(/react/i)
 // learn more: https://github.com/testing-library/jest-dom
 import '@testing-library/jest-dom/extend-expect'
+import ResizeObserver from 'resize-observer-polyfill'
+global.ResizeObserver = ResizeObserver
```
Comment (thisislawatts) [inline src/setupTests.ts:7]: **suggestion:** It would be worth explicitly adding this as a dev dependency, otherwise there is a risk it could be removed accidentally.

`yarn add --dev resize-observer-polyfill`

Comment (AlfonsoGhislieri) [reply src/setupTests.ts:7]: Good idea! Added in 8b1b0e8

**Comment by thisislawatts** [comment]
> This is PART 1, a lot of this functionality repeats with what already exists for the howtos. In order to avoid bloating this PR too much PART 2 will address refactoring, finally PART 3 will add icrementing the downloads via a firebase function to avoid unnecessary database calls.

Would it be worth adding these steps as a task list to the original issue?

**Comment by AlfonsoGhislieri** [comment]
Good idea @thisislawatts, just added them now! Also this is ready to review, sorry for all the commits and changes I maybe should have already broken up this PR more.. 😬

**Comment by thisislawatts** [comment]
**question:** Did you take a look at what tests we could introduce here? I am thinking about the misbehaviour [[bug] Upload file not working in howto](https://github.com/ONEARMY/community-platform/issues/2244) you have found and whether we should prioritise addressing that before moving forward with this PR.

My reasoning is that:
* We should prioritise addressing functional defects in production before new features
* Any testing we introduce for how-to downloads can be adopted for Research downloads.

File: src/models/research.models.tsx [hunk 2]
```diff
@@ -24,6 +24,9 @@ export namespace IResearch {
     total_views?: number
     collaborators: string[]
     subscribers?: UserIdList
+    total_downloads?: number
+    fileLink?: string
+    files?: Array<IUploadedFileMeta | File | null>
```
Comment (thisislawatts) [inline src/models/research.models.tsx:29]: **suggestion:** Would you be up for pushing these down under a common "namespace". The intention here being to remove some noise from the object root and group related properties together. 

Something like: 

```ts
type Item = {
  ...
  attachments?: {
    total_downloads: number;
    fileLink?: string;
    files?:  Array<blah>
  }
}
```

File: src/pages/Howto/Content/Howto/HowtoDescription/HowtoDescription.tsx [hunk 3]
```diff
@@ -290,9 +290,9 @@ const HowtoDescription = ({ howto, loggedInUser, ...props }: IProps) => {
               <Text
                 data-cy="file-download-counter"
                 sx={{
-                  fontSize: '12px',
-                  color: '#61646B',
-                  paddingLeft: '8px',
```
Comment (thisislawatts) [inline src/pages/Howto/Content/Howto/HowtoDescription/HowtoDescription.tsx:295]: **praise:** Great to see this improvement, and happy to wave it through here but really it is not related to Research module so could be split into a separate PR.

File: src/pages/Research/Content/Common/Research.form.tsx [hunk 4]
```diff
@@ -13,25 +13,29 @@ import {
...
 import { COMPARISONS } from 'src/utils/comparisons'
 import { stripSpecialCharacters } from 'src/utils/helpers'
-import { required } from 'src/utils/validators'
+import { required, validateUrlAcceptEmpty } from 'src/utils/validators'
 import { PostingGuidelines } from './PostingGuidelines'
 import { ResearchSubmitStatus } from './SubmitStatus'
 import { CategoriesSelect } from 'src/pages/Howto/Category/CategoriesSelect'
 import { RESEARCH_TITLE_MAX_LENGTH, RESEARCH_MAX_LENGTH } from '../../constants'
 
+const MAX_LINK_LENGTH = 2000
```
Comment (thisislawatts) [inline src/pages/Research/Content/Common/Research.form.tsx:30]: **nitpick(non-blocking):** Move this to one of the `constants.ts` files as it is probably a constraint shared with the how-to files. Non-blocking as it may be something you had already identified for refactoring.

**Comment by davehakkens** [comment]
Nice. thanks for adding!
Originally the issue was aiming for each step/update to option to attach files. Just as they can add videos or images.
Mainly because reseach is an ongoing proces, when you start you don't have the files yet. You will attach them as you move forward researching

I can imagine this makes it more complex since it's a bit different from the original usage in how-to. 
Your PR could also work for now, but i'd  be curious to hear what would be preffered from a dev point of view.

File: src/models/research.models.tsx (hunk 2 above)
Comment (AlfonsoGhislieri) [reply src/models/research.models.tsx]: Really like this idea, great suggestion! Added in: 0e6dcc9. Will also extend this to the howto's when I get to the refactoring!

File: src/pages/Research/Content/Common/Research.form.tsx (hunk 4 above)
Comment (AlfonsoGhislieri) [reply src/pages/Research/Content/Common/Research.form.tsx:30]: Yes! I have this on the TODO for the refactoring that comes next.

**Comment by AlfonsoGhislieri** [comment]
Ah I see @davehakkens, I didn't know we were going to extend it for updates but that makes sense. Considering there is still a fair bit of work to go to get this to be less messy I added it to the issue TODO list and this will be addressed in Part 4 once I have refactored and implemented firebase functions.

@thisislawatts I agree I think we should freeze for now and I will look into adding some better test coverage that we can then extend to the research articles.

Will mark this as blocked until this is in place.

**Comment by AlfonsoGhislieri** [comment]
#2280 Has now been added to unblock #2244, this PR will also need to be updated with the suitable logic and tests to give feedback when a user tries to add both a link and uploads files. This again just really points towards how important it would be to find a way to combine both howto and research to avoid all this duplication of work and code :(

**Comment by AlfonsoGhislieri** [comment]
Closing this as its become very stale and I’m working on a different implementation
